File: ios/music/music/Song.swift
import Foundation

struct PlaylistEntry: Codable {
    var name: String
    var imageUrl: String?
    var contains: Bool

    enum CodingKeys: String, CodingKey {
        case name, contains
        case imageUrl = "image_url"
    }
}

class Song: Codable, Identifiable {
    var isrc: String
    var title: String
    var duration: Int
    var imageUrl: String?
    var artist: String?
    var album: String?
    var isInPlaylistMap: [String: PlaylistEntry]?

    enum CodingKeys: String, CodingKey {
        case isrc, title, artist, album, duration
        case imageUrl = "image_url"
        case isInPlaylistMap = "is_in_playlist_map"
    }

    init(isrc: String = "", title: String = "", duration: Int = 0, imageUrl: String? = nil, artist: String? = nil, album: String? = nil) {
        self.isrc = isrc
        self.title = title
        self.duration = duration
        self.imageUrl = imageUrl
        self.artist = artist
        self.album = album
    }
}

struct SearchResults: Codable {
    var songs: [Song]
    var playlists: [Playlist]
}

File: app/Models/Song.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Song extends Model
{

    protected $primaryKey = 'isrc';
    public $incrementing = false;
    protected $keyType = 'string';

    protected $fillable = [
        'title',
        'artist',
        'album',
        'image_url',
        'duration',
        'isrc',
    ];

    public function playlists()
    {
        return $this->belongsToMany(Playlist::class, 'playlist_songs', 'song_isrc', 'playlist_id', 'isrc', 'id')->withTimestamps();
    }
}
